service/importer/definition.go

```diff
@@ -1,108 +1,111 @@
+	Dependencies map[string]*Dependency `yaml:"dependencies" json:"dependencies,omitempty" validate:"dive,keys,printascii,ne=service,endkeys,required"`
 
 	// Configuration is the Docker container that service runs inside.
 	Configuration *Dependency `yaml:"configuration" json:"configuration,omitempty"`
 
 	// Repository holds the service's repository url if it's living on
 	// a Git host.
-	Repository string `yaml:"repository" json:"repository,omitempty"`
+	Repository string `yaml:"repository" json:"repository,omitempty" validate:"omitempty,uri"`
 }
 
 // Event describes a service task.
 type Event struct {
 	// Name is the name of event.
-	Name string `yaml:"name" json:"name,omitempty"`
+	Name string `yaml:"name" json:"name,omitempty" validate:"printascii"`
 
 	// Description is the description of event.
-	Description string `yaml:"description" json:"description,omitempty"`
+	Description string `yaml:"description" json:"description,omitempty" validate:"printascii"`
 
 	// Data holds the input parameters of event.
-	Data map[string]*Parameter `yaml:"data" json:"data,omitempty"`
+	Data map[string]*Parameter `yaml:"data" json:"data,omitempty" validate:"required,dive,keys,printascii,endkeys,required"`
 }
 
 // Dependency represents a Docker container and it holds instructions about
 // how it should run.
 type Dependency struct {
 	// Image is the Docker image.
-	Image string `yaml:"image" json:"image,omitempty"`
+	Image string `yaml:"image" json:"image,omitempty" validate:"printascii"`
 
 	// Volumes are the Docker volumes.
-	Volumes []string `yaml:"volumes" json:"volumes,omitempty"`
+	Volumes []string `yaml:"volumes" json:"volumes,omitempty" validate:"unique,dive,printascii"`
 
 	// VolumesFrom are the docker volumes-from from.
-	VolumesFrom []string `yaml:"volumesfrom" json:"volumefrom,omitempty"`
+	VolumesFrom []string `yaml:"volumesfrom" json:"volumefrom,omitempty" validate:"unique,dive,printascii"`
 
 	// Ports holds ports configuration for container.
-	Ports []string `yaml:"ports" json:"ports,omitempty"`
+	Ports []string `yaml:"ports" json:"ports,omitempty" validate:"unique,dive,numeric"`
 
 	// Command is the Docker command which will be executed when container started.
-	Command string `yaml:"command" json:"command,omitempty"`
+	Command string `yaml:"command" json:"command,omitempty" validate:"printascii"`
 
 	// Args hold the args to pass to the Docker container
-	Args []string `yaml:"args" json:"args,omitempty"`
+	Args []string `yaml:"args" json:"args,omitempty" validate:"dive,printascii"`
 
 	// Env is a slice of environment variables in key=value format.
-	Env []string `yaml:"env" json:"env,omitempty"`
+	Env []string `yaml:"env" json:"env,omitempty" validate:"unique,dive,printascii"`
 }
 
 // Task describes a service task.
 type Task struct {
 	// Name is the name of task.
-	Name string `yaml:"name" json:"name,omitempty"`
+	Name string `yaml:"name" json:"name,omitempty" validate:"printascii"`
```

Comment:
```suggestion
	Name string `yaml:"name" json:"name,omitempty" validate:"printascii,min=1"`
```

Reply:
Same here, `Name` is not required, the key in the map is